lib/taskana-core/src/main/java/pro/taskana/impl/TaskServiceImpl.java

```diff
@@ -759,29 +680,186 @@ private void standardSettings(TaskImpl task, Classification classification,
+                bulkLog.addError(currentTaskId,
+                    new NotAuthorizedException(
+                        "The workbasket of this task got not TRANSFER permissions. TaskId=" + currentTaskId));
+                taskIdIterator.remove();
+            }
+        }
+    }
+
+    private void checkIfTasksMatchCompleteCriteria(List<String> taskIds, List<TaskSummary> taskSummaries,
+        BulkOperationResults<String, TaskanaException> bulkLog) {
+        Instant now = Instant.now();
+        Iterator<String> taskIdIterator = taskIds.iterator();
+        while (taskIdIterator.hasNext()) {
+            String currentTaskId = taskIdIterator.next();
+            TaskSummaryImpl taskSummary = (TaskSummaryImpl) taskSummaries.stream()
+                .filter(ts -> currentTaskId.equals(ts.getTaskId()))
+                .findFirst()
+                .orElse(null);
+            if (taskSummary == null) {
+                bulkLog.addError(currentTaskId, new TaskNotFoundException(currentTaskId, "task with id "
+                    + currentTaskId + " was not found."));
+                taskIdIterator.remove();
+            } else if (taskSummary.getClaimed() == null || taskSummary.getState() != TaskState.CLAIMED) {
+                bulkLog.addError(currentTaskId, new InvalidStateException(currentTaskId));
+                taskIdIterator.remove();
+            } else if (!CurrentUserContext.getAccessIds().contains(taskSummary.getOwner())) {
+                bulkLog.addError(currentTaskId, new InvalidOwnerException(
+                    "TaskOwner is" + taskSummary.getOwner() + ", but current User is "
+                        + CurrentUserContext.getUserid()));
+                taskIdIterator.remove();
+            } else {
+                taskSummary.setCompleted(now);
+                taskSummary.setModified(now);
+                taskSummary.setState(TaskState.COMPLETED);
+            }
+        }
+    }
+
+    private void updateTasksToBeTransferred(List<String> taskIds,
+        List<MinimalTaskSummary> taskSummaries, Workbasket destinationWorkbasket) {
+
+        taskSummaries = taskSummaries.stream()
+            .filter(ts -> taskIds.contains(ts.getTaskId()))
+            .collect(
+                Collectors.toList());
+        if (!taskSummaries.isEmpty()) {
+            Instant now = Instant.now();
+            TaskSummaryImpl updateObject = new TaskSummaryImpl();
+            updateObject.setRead(false);
+            updateObject.setTransferred(true);
+            updateObject.setWorkbasketSummary(destinationWorkbasket.asSummary());
+            updateObject.setDomain(destinationWorkbasket.getDomain());
+            updateObject.setModified(now);
+            updateObject.setState(TaskState.READY);
+            updateObject.setOwner(null);
+            taskMapper.updateTransfered(taskIds, updateObject);
+            createTasksTransferredEvent(taskSummaries, updateObject);
```

Comment:
Since this is a complex operation, wouldn't it be good to have a check like if (HistoryEventProducer.isHistoryEnabled()) here as well?